### PR TITLE
Don't use write barriers for byref-like struct copies

### PR DESCRIPTION
**Pull request by EgorBo**
Byref-like structs are stack only so we can omit write barriers for them. It looks like JIT already knew that, but only if such structs had `ref` fields, which they might not have.

```cs
void Test(ref MyStruct dst, MyStruct src) => dst = src;

ref struct MyStruct
{
    object a;
    long b;
}
```
Current codegen for `Test`:
```asm
; Method ConsoleApp35.Program:Test(byref,ConsoleApp35.MyStruct):this (FullOpts)
       push     rdi
       push     rsi
       mov      rdi, rdx
       mov      rsi, r8
       call     CORINFO_HELP_ASSIGN_BYREF
       movsq    
       pop      rsi
       pop      rdi
       ret      
; Total bytes of code: 18
```
New codegen for `Test`:
```asm
; Method ConsoleApp35.Program:Test(byref,ConsoleApp35.MyStruct):this (FullOpts)
       vmovdqu  xmm0, xmmword ptr [r8]         ;; NOTE: nogc region so SIMD is fine
       vmovdqu  xmmword ptr [rdx], xmm0
       ret      
; Total bytes of code: 10
```

### CODE REVIEW

**Comment by EgorBo** [comment]
@MihuBot

**Comment by EgorBo** [comment]
@EgorBot
```cs
using System.Runtime.CompilerServices;
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;

BenchmarkRunner.Run<WbBench>(args: args);

public class WbBench
{
    [Benchmark]
    public void Bench()
    {
        MyStruct a = new();
        MyStruct b = new();
        CopyBlock(ref a, b);
    }

    [MethodImpl(MethodImplOptions.NoInlining)]
    void CopyBlock(ref MyStruct dst, MyStruct src) => dst = src;

    ref struct MyStruct
    {
        object a;
        long b;
    }
}
```

File: src/coreclr/jit/lowerarmarch.cpp [hunk 1]
```diff
@@ -735,7 +735,7 @@ void Lowering::LowerBlockStore(GenTreeBlk* blkNode)
         {
             // No write barriers are needed on the stack.
             // If the layout contains a byref, then we know it must live on the stack.
```
Comment (teo-tsirpanis) [inline src/coreclr/jit/lowerarmarch.cpp:737]: Update comments?

Comment (EgorBo) [reply src/coreclr/jit/lowerarmarch.cpp:737]: Thanks! fixed

**Comment by EgorBo** [comment]
PTAL @jakobbotsch @kunalspathak @dotnet/jit-contrib [diffs are relatively small](https://dev.azure.com/dnceng-public/public/_build/results?buildId=678161&view=ms.vss-build-web.run-extensions-tab) since there are not so many byref structs without ref fields.

**Review by kunalspathak — APPROVED** [review]
LGTM

File: src/coreclr/jit/layout.cpp [hunk 2]
```diff
@@ -421,22 +421,22 @@ void ClassLayout::InitializeGCPtrs(Compiler* compiler)
...
 // Return value:
-//    true if at least one GC ByRef, false otherwise.
+//    true if the block is stack only
 //
-bool ClassLayout::HasGCByRef() const
+bool ClassLayout::IsStackOnly(Compiler* comp) const
 {
-    unsigned slots = GetSlotCount();
-    for (unsigned i = 0; i < slots; i++)
+    // Byref-like structs are stack only
```
Comment (kunalspathak) [inline src/coreclr/jit/layout.cpp:434]: Is there a documentation that I can read more about this?

**Review by jakobbotsch — APPROVED** [review]

File: src/coreclr/jit/layout.cpp (hunk 2 above)
Comment (EgorBo) [reply src/coreclr/jit/layout.cpp:434]: Yep, e.g. https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/ref-struct Basically, Span itself is the most popular example

**Pull request merged**